azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py

```diff
@@ -0,0 +1,112 @@
+import copy
+import itertools
+import json
+import os
+from typing import List, Iterable
+from urllib.parse import urlparse
+
+from azure.kusto.data.helpers import get_string_tail_lower_case
+from azure.kusto.data.security import _is_local_address
+
+from .exceptions import KustoClientInvalidConnectionStringException
+
+
+class MatchRule:
+    def __init__(self, suffix, exact):
+        self.suffix = suffix.lower()
+        self.exact = exact
+
+
+class FastSuffixMatcher:
+    def __init__(self, rules: Iterable[MatchRule]):
+        it1, it2 = itertools.tee(rules)
+        self._suffix_length = min([len(rule.suffix) for rule in it1])
+        _processed_rules = {}
+        for rule in it2:
+            suffix = get_string_tail_lower_case(rule.suffix, self._suffix_length)
+            if suffix not in _processed_rules:
+                _processed_rules[suffix] = []
+            _processed_rules[suffix].append(rule)
+
+        self.rules = _processed_rules
+
+    def is_match(self, candidate):
+        if len(candidate) < self._suffix_length:
+            return False
+
+        _match_rules = self.rules.get(get_string_tail_lower_case(candidate, self._suffix_length))
+        if _match_rules:
+            for rule in _match_rules:
+                if candidate.lower().endswith(rule.suffix):
+                    if len(candidate) == len(rule.suffix) or not rule.exact:
+                        return True
+
+        return False
+
+
+def create_fast_suffix_matcher_from_existing(rules: List[MatchRule], existing: FastSuffixMatcher) -> FastSuffixMatcher:
+    if existing is None or len(existing.rules) == 0:
+        return FastSuffixMatcher(rules)
+
+    if not rules:
+        return existing
+
+    lst = copy.deepcopy(rules)
+    for _, v in existing.rules.items():
+        lst = lst + v
+    return FastSuffixMatcher(lst)
+
+
+class KustoTrustedEndpoints:
+    def __init__(self):
+        self._matchers = dict()
+        for k, v in _well_known_kusto_endpoints_data["AllowedEndpointsByLogin"].items():
+            rules = itertools.chain(
```

Comment:
Iterators are a bit overkill for a small list of known size. If it's just to combine them, use splats:

```python
[*(MatchRule(suffix, False) for suffix in v["AllowedKustoSuffixes"]), *(MatchRule(hostname, True) for hostname in v["AllowedKustoHostnames"])]
```

Reply:
thanks its so much better